fleetd: keep a list of listeners across reconfiguring fleetd

dongsupark opened the pull request:
So far fleetd has lost its existing listeners after receiving SIGHUP.
That happened because ``reconfigure()`` simply killed the server and started
a fresh one, without taking consideration into the existing listeners.
That resulted in a strange error like:

>  Failed serving HTTP on listener: @ accept unix @: accept: invalid argument

So let's keep the list of existing listeners before killing the server,
and give it to the new server.

Fixes: https://github.com/coreos/fleet/issues/1190

/cc @kayrus @dalbani

antrik commented on server/server.go:
```diff
@@ -115,9 +116,17 @@ func New(cfg config.Config) (*Server, error) {
+		listeners = make([]net.Listener, len(oldListeners))
+		for i, l := range oldListeners {
+			listeners[i] = l
+		}
```
Why are you copying the values one by one: can't you just assign `listeners = oldListeners`?...

(This will just copy the reference rather than the actual values -- but I don't see why this would be undesirable?)

dongsupark replied:
@antrik 
> can't you just assign listeners = oldListeners?

Correct. Done.

antrik commented:
Looks good aside from the one patch comment. I'd say this should come with a testcase, though?

dongsupark commented:
@antrik 
> I'd say this should come with a testcase, though?

Done. I added a new functional test ``TestReconfigureServer``.

jonboulle commented on server/server.go:
```diff
@@ -115,9 +116,14 @@ func New(cfg config.Config) (*Server, error) {
-	listeners, err := activation.Listeners(false)
-	if err != nil {
-		return nil, err
+	var listeners []net.Listener
```
A little simpler
`func New(cfg config.Config, listeners []net.Listener) (*Server, error) {`
```
if len(listeners) == 0 {
    listeners, err = activation.Listeners ...
}
```

jonboulle replied:
== nil check is unnecessary, len handles it for you

dongsupark replied:
@jonboulle It's done. Thanks for the suggestion.

jonboulle commented:
Almost LGTM but perhaps consider https://github.com/coreos/fleet/pull/1528/files ?

dongsupark commented:
@jonboulle Done. I dropped unnecessary sudo calls, except for ``sudo kill -HUP``.

tixxdz commented on functional/server_test.go:
```diff
@@ -0,0 +1,78 @@
+	// Without this sleep, the entire fleetd test always ends up succeeding
+	// no matter whether SIGHUP came or not.
+	_, _ = cluster.MemberCommand(m0, "sh", "-c", `'sleep 2'`)
+
```
Hmm, would it help after SIGHUP is sent to check the journal if "fleet.*Reloading configuration.*" is really there ?

antrik replied:
AIUI the problem is actually that we need to make sure fleetd is fully initialised before even sending the SIGHUP. I agree though that it would be better to wait for this by checking the logs. (Unless there is some more elegant method for checking fleetd's exact status?...)

Note that a proper check like that will also make the preceding checks using systemctl unnecessary.

dongsupark replied:
@tixxdz 
> Hmm, would it help after SIGHUP is sent to check the journal if "fleet.Reloading configuration." is really there ?

Yes, I updated it like that.
But that change didn't allow me to remove the sleep command before sending SIGHUP. Without the sleep, test always ends up succeeding, no matter what.. :-(

antrik replied:
@dongsupark actually, I just realised that `WaitForNMachines()` will only succeed once fleet is fully up? So the `systemctl` checks should definitely be redundant.

Also, I still don't understand why the sleep is necessary... But I guess that's not a showstopper.

tixxdz commented on functional/server_test.go:
```diff
@@ -0,0 +1,78 @@
+	if strings.TrimSpace(stdout) != "" {
+		t.Fatalf("Fleetd log returned error on HTTP listeners: %s", stdout)
+	}
+
```
The check seems fine, but what worries me is we are checking an error message that might disappear for other reasons! is there a better way to check for sockets in: systemctl status fleetd; grep pid, then inspect /proc/pid/* ?

dongsupark replied:
Yeah, I was just lazy when I wrote this part. I'll try to change it like that.

antrik replied:
Trying to check sockets sounds complicated, and I'm not convinced it's really a good idea to rely on internals like that...

TBH, I don't think *either* of these approaches is really the right way to test this. Rather, we should try to check whether actual socket connections keep working as expected. I don't have any specific setup in mind for this -- but surely it should be possible to do this?

tixxdz replied:
@antrik indeed, if it's easy to check if the socket is still working in this context then go for it ? my comment was a bit generic...

jonboulle replied:
> Rather, we should try to check whether actual socket connections keep working as expected.

+1, test for the desired behaviour by testing the desired behaviour :-)

dongsupark replied:
> we should try to check whether actual socket connections keep working as expected

Done.

tixxdz commented:
I just commented, if there are no easy ways! then don't bother, lgtm.

Thank you!

antrik commented on functional/server_test.go:
```diff
@@ -0,0 +1,137 @@
+	err = waitForFleetdSocket(cluster, m0)
+	if err != nil {
+		t.Fatalf("Failed to get a list of fleetd sockets: %v", err)
+	}
```
Do we still need this? I would assume it's no longer necessary, now that we actually check for the socket connection still working?...

antrik replied:
I see that you use it here now to check for fleet being up, which makes sense I guess. With this in place, do you still need the `sleep` above?...

dongsupark replied:
> I see that you use it here now to check for fleet being up, which makes sense I guess. With this in place, do you still need the sleep above?...

Unfortunately yes, I still need the sleep. I just did another round of tests, but it's still the same.
As mentioned above, without the sleep, the whole test always succeeds. What would be the point testing this at all, if it would always give the same answer? That's where I'm puzzled, but for me it's more important to get the test working at all.

dongsupark commented:
Updated.
* remove unnecessary checks for systemctl show in the beginning
* remove unnecessary checks for fleetd socket after sending the signal
* split out ``waitForReloadConfig`` for better readability

tixxdz commented on functional/server_test.go:
```diff
@@ -0,0 +1,133 @@
+			// in a subtle way, so that it sometimes fails only on semaphoreci.
+			// - dpark 20160408
+			stdout, _ := cluster.MemberCommand(m0, "journalctl -u fleet")
+			journalfleet := strings.TrimSpace(stdout)
```
@dongsupark does it make a difference here if you do "journalctl _PID=$(pidof fleetd)" ?

antrik replied:
I don't see why `pidof` would be more reliable than `-u`? Do we expect more than one instance of fleetd to exist during the test? (And if so, is it actually appropriate to consider only the currently active one?...)

(I don't see why using grep wouldn't be reliable in the first place, though... But doing the compare in Go is fine too; so I guess this is a non-issue.)

tixxdz replied:
What are the gains of using `-u` where we just want what the pid logs which is more precise in this context ?

tixxdz replied:
and yeh I forget to say that some developers are also just lazy so they run the tests again and again on the same VM grepping the old bogus journal logs...

antrik replied:
@tixxdz there are no "old bogus journal logs" -- each test spawns fresh containers with fresh logs. And I still think the `-u` abstraction offered to us by `journalctl` is more elegant than dealing with low-level details such as PIDs ourselves...

(PIDs are in fact prone to race conditions in general -- though I guess that's not really an issue in this specific case.)

tixxdz commented on functional/server_test.go:
```diff
@@ -0,0 +1,133 @@
+
+	// check if fleetd is still running correctly, by running fleetctl status
+	_, _, err = cluster.Fleetctl(m0, "status")
+	if err != nil {
```
@dongsu: could you please change this to:
`fleet $ ./bin/fleetctl list-units; echo $?
Error retrieving list of units from repository: googleapi: Error 503: fleet server unable to communicate with etcd
1
`
`fleet $ ./bin/fleetctl status; echo $?
0
`
That's my result of experimenting right now, at least it is the same as in the original reported bug

Thanks you!

dongsupark replied:
@tixxdz this reveals that there's still a bug somewhere. I'll try to investigate soon.

dongsupark replied:
@tixxdz Fixed a bug. Checking with "fleetctl list-units" is now working.

tixxdz commented:
Lgtm, and since this is fixing a bug that most users want I'm merging this series. If any one have a problem with that minor sleep in the functional test feel free to open an issue or a PR. Sure it helps to serialize things when the container and all the services are started and where the logs and extra operation may have some delays...

Thanks!
